New Branch for App-2

abubakar-a1qa opened the pull request.

abubakar-a1qa committed:
Cleared the branch

abubakar-a1qa committed:
Updated the gitignore

abubakar-a1qa committed:
Integrated 2nd test

abubakar-a1qa committed:
updated gitIgnore

abubakar-a1qa committed:
delete mochawesome

abubakar-a1qa committed:
added mochawesome-report directory to gitignore

abubakar-a1qa committed:
Integrated both tests and working fine

MarkAtolog commented on pageObjects/CommunityMarketPage.js:
```diff
@@ -0,0 +1,111 @@
+        const priceTags = await this.driver.findElements(this.locators.priceTagLocator);
+        let sorted = true;
+
+        for (let i = 0; i < priceTags.length - 1; i++) {
+            const price1Text = await priceTags[i].getText();
+            const price2Text = await priceTags[i + 1].getText();
+
+            // Remove non-numeric characters and parse as a float
+            const price1 = parseFloat(price1Text.replace(/[^0-9.-]+/g, ''));
+            const price2 = parseFloat(price2Text.replace(/[^0-9.-]+/g, ''));
+
+            // Check if prices are not in ascending order and not equal
+            if (price1 > price2 && price1 !== price2) {
+                sorted = false;
+                break;
+            }
+        }
+        return sorted;
+    }
+
+    async isPriceSortedInDescendingOrder() {
+        const priceTags = await this.driver.findElements(this.locators.priceTagLocator);
+        let sorted = true;
+
+        for (let i = 0; i < priceTags.length - 1; i++) {
+            const price1Text = await priceTags[i].getText();
+            const price2Text = await priceTags[i + 1].getText();
+
+            // Remove non-numeric characters and parse as a float
+            const price1 = parseFloat(price1Text.replace(/[^0-9.-]+/g, ''));
+            const price2 = parseFloat(price2Text.replace(/[^0-9.-]+/g, ''));
+
+            // Check if prices are not in descending order and not equal
+            if (price1 < price2 && price1 !== price2) {
+                sorted = false;
+                break;
+            }
+        }
+        return sorted;
+    }
```
what do you think is percent of code duplication here? is it a good practice to write so much duplicated code?

MarkAtolog commented on steps/CommunityMarketSteps.js:
```diff
@@ -0,0 +1,46 @@
+
+    async openAdvancedSearchOptions() {
+        await this.communityMarketPage.clickOnShowAdvancedOptionsDropdown();
+        return await this.communityMarketPage.isShowAdvancedOptionWindowVisible();
```
don't forget about "return await"

abubakar-a1qa committed:
Updated the code to reduce wait time.

abubakar-a1qa committed:
Cleaning unnecessary timeouts

MarkAtolog commented on pageObjects/CommunityMarketPage.js:
```diff
@@ -61,10 +62,34 @@ class CommunityMarketPage extends BasePage {
+            if (isChanged) {
+                return true;
+            }
+        }
```
possible infinite loop

MarkAtolog commented on pageObjects/CommunityMarketPage.js:
```diff
@@ -0,0 +1,136 @@
+const BasePage = require('./BasePage');
+const CommunityMarketPageLocators = require('../locators/CommunityMarketPageLocators');
+const { By } = require('selenium-webdriver');
+const assert = require('assert');
+const TimeOut = require("../utils/timeouts/TimeOut");
```
imports should be sorted (external libraries first)

MarkAtolog commented on pageObjects/CommunityMarketPage.js:
```diff
@@ -0,0 +1,136 @@
+        await this.click(this.locators.searchButtonLocator);
+    }
+
+    async isFilterApplied(filterItemNames) {
+        for (const filterItemName of filterItemNames) {
+            const filterLocator = By.xpath(this.locators.appliedFilterItemsLocator(filterItemName));
+            const filterElement = await this.driver.findElement(filterLocator);
+            const filterText = await filterElement.getText();
+            assert(filterText.includes(filterItemName));
+        }
+    }
```
page can't assert anything. Only actions

MarkAtolog commented on steps/CommunityMarketSteps.js:
```diff
@@ -0,0 +1,54 @@
+
+class CommunityMarketSteps {
+    constructor(driver) {
+        this.communityMarketPage = new (require('../pageObjects/CommunityMarketPage'))(driver);
```
all imports should be located in the same place at the start of the file

abubakar-a1qa committed:
Implemented last task provided by mentor and fixed the comments

abubakar-a1qa committed:
removed await which I placed after return

abubakar-a1qa committed:
updated gitignore

MarkAtolog commented on pageObjects/CommunityMarketPage.js:
```diff
@@ -75,61 +78,52 @@ class CommunityMarketPage extends BasePage {
-        for (let i = 0; i < priceTags.length - 1; i++) {
-            const price1Text = await priceTags[i].getText();
-            const price2Text = await priceTags[i + 1].getText();
+    async waitForItemsToChange(resultList, maxRetries = 3, intervalTime = TimeOut.OneSecond) {
```
one second is too much

MarkAtolog commented on pageObjects/CommunityMarketPage.js:
```diff
@@ -75,61 +78,52 @@ class CommunityMarketPage extends BasePage {
-                break;
+            if (!isChanged) {
+                retries++;
+                await this.driver.sleep(intervalTime);
```
don't use sleep in your code, where it's not necessary (almost everywhere it's unnecessary)

MarkAtolog commented on pageObjects/CommunityMarketPage.js:
```diff
@@ -75,61 +78,52 @@ class CommunityMarketPage extends BasePage {
-            // Remove non-numeric characters and parse as a float
-            const price1 = parseFloat(price1Text.replace(/[^0-9.-]+/g, ''));
-            const price2 = parseFloat(price2Text.replace(/[^0-9.-]+/g, ''));
+        while (!isChanged && retries < maxRetries) {
+            const updatedResultList = await this.getItemNamesFromResultList();
```
use driver.wait() instead of loop

MarkAtolog commented on steps/CommunityMarketSteps.js:
```diff
@@ -18,10 +19,20 @@ class CommunityMarketSteps {
-        await this.communityMarketPage.isFilterApplied(filterItems);
+        for (const filterItem of filterItems) {
+            const isApplied = await this.communityMarketPage.isFilterApplied(filterItem);
+            assert(isApplied, `Filter ${filterItem} is not applied`);
```
change 'assert' library to asserts from 'chai' library

MarkAtolog commented on .idea/workspace.xml:
```diff
@@ -4,16 +4,31 @@
-    <list default="true" id="d86f92bd-42e5-419e-8362-e5a9289bc8cf" name="Changes" comment="vcs.xml" />
+    <list default="true" id="88e23f2f-48dd-46bb-9f10-6feb402ad457" name="Changes" comment="Cleaning unnecessary timeouts">
+      <change afterPath="$PROJECT_DIR$/resources/testData.json" afterDir="false" />
+      <change beforePath="$PROJECT_DIR$/.idea/workspace.xml" beforeDir="false" afterPath="$PROJECT_DIR$/.idea/workspace.xml" afterDir="false" />
```
do not commit this file!
file should be removed

MarkAtolog commented on pageObjects/CommunityMarketPage.js:
```diff
@@ -0,0 +1,130 @@
+
+    async getItemNamesFromResultList() {
+        const itemNames = [];
+        for (let i = 0; i <= 9; i++) {
```
magic number (9). What is the 9 here?

MarkAtolog commented on pageObjects/CommunityMarketPage.js:
```diff
@@ -0,0 +1,130 @@
+        }
+    }
+
+    /*async waitForItemsToChange(resultList) {
+        let isChanged = false;
+
+        while (!isChanged) {
+            const updatedResultList = await this.getItemNamesFromResultList();
+            isChanged = resultList.some((item, index) => item !== updatedResultList[index]);
+        }
+        return isChanged;
+    }*/
```
remove commented code

MarkAtolog commented on tests/communityMarketTest.js:
```diff
@@ -0,0 +1,91 @@
+        communityMarketSteps = new CommunityMarketSteps(driver);
+    });
+
+    it('Verify if 1st selected item is Displayed', async function () {
+        await homePageSteps.navigateToCommunityMarket();
+        const currentUrl = await driver.getCurrentUrl();
+        assert(currentUrl.includes('market'), 'Not on the Community Market page');
+
+        const isAdvancedOptionsVisible = await communityMarketSteps.openAdvancedSearchOptions();
+        assert(isAdvancedOptionsVisible, 'Advanced options window is not visible');
+
+        await communityMarketSteps.searchForItem('Dota 2', 'Phantom Assassin', 'Rare');
+        await communityMarketSteps.validateFilters(['Dota 2', 'Phantom Assassin', 'Rare']);
+
+        const isResultValid = await communityMarketSteps.verifySearchResult(0);
+        assert(isResultValid, 'Item title on the page does not match the result text');
+    });
```
why this scenario is not added to json file with scenarios?

abubakar-a1qa replied:
Because I wanted to keep this one separate, so that I can keep track of my work. If you need them, I can update the json file and add these data on the tdd case.

MarkAtolog replied:
there is no reason to keep, there is no difference with another test, it's completely the same, just different test data. If you want to track you work - you have commits history, everything you do is stored in commits, you can open any commit at any time

abubakar-a1qa committed:
Updated the code according to the mentor's comment. Used Chai, Updated timeout etc.

Updated isMarketTableListingMessageVisible(), validateFilters(), waitForItemsToChange()

MarkAtolog commented on .idea/workspace.xml:
```diff
@@ -4,14 +4,11 @@
   </component>
   <component name="ChangeListManager">
-    <list default="true" id="88e23f2f-48dd-46bb-9f10-6feb402ad457" name="Changes" comment="Cleaning unnecessary timeouts">
-      <change afterPath="$PROJECT_DIR$/resources/testData.json" afterDir="false" />
+    <list default="true" id="88e23f2f-48dd-46bb-9f10-6feb402ad457" name="Changes" comment="">
```
and again you pushed this file. Do you know how to select only necessary file for commit?

MarkAtolog replied:
Don't add all files in index for commit. select only necessary file, add them to index and commit. Now just delete this file from the branch, it should not be there

MarkAtolog commented on pageObjects/CommunityMarketPage.js:
```diff
@@ -67,9 +66,20 @@ class CommunityMarketPage extends BasePage {
+
+            const element = await this.driver.findElements(itemLocator);
+
+            if (element.length === 0) {
+                break;
+            }
```
put this condition inside while()
While (true) is a bad practice even if you have break inside. Use correct condition
Previous for loop was better, just don't use hardcoded and magic numbers

MarkAtolog replied:
just put in for loop correct number. If you are iterating by lines, then you need to get count of that lines. How to get count of the lines? use findElement function and you will receive array of elements, get length, and this will be number you can put in for loop. Or you can just iterate by elements in array

MarkAtolog commented on pageObjects/CommunityMarketPage.js:
```diff
@@ -87,31 +97,11 @@ class CommunityMarketPage extends BasePage {
-        let retries = 0;
-
-        while (!isChanged && retries < maxRetries) {
+    async waitForItemsToChange(resultList, timeout = TimeOut.HalfSecond) {
```
are you sure items will change in half of the second. Timeout is too narrow, give more time

MarkAtolog replied:
half second could be ok for interval, in some cases, but definitely not enough for timeout.

MarkAtolog commented on pageObjects/CommunityMarketPage.js:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    async waitForItemsToChange(resultList, timeout = TimeOut.HalfSecond) {
+        return await this.driver.wait(async () => {
```
return await?

MarkAtolog replied:
we discussed it several times, it's redundant. This code is similar to "await return await", where first await is for waitForItemsToChange function

MarkAtolog commented on tests/communityMarketTest.js:
```diff
@@ -0,0 +1,91 @@
+const { Builder } = require('selenium-webdriver');
+const chrome = require('selenium-webdriver/chrome');
+require('chromedriver');
+const assert = require('assert');
```
assertion library is not changed

MarkAtolog replied:
if comment left in one file - it still means, that you need to fix that comment everywhere

MarkAtolog commented on tests/communityMarketTest.js:
```diff
@@ -0,0 +1,91 @@
+        let options = new chrome.Options().addArguments('--start-maximized');
+        driver = new Builder().forBrowser('chrome').setChromeOptions(options).build();
+
+        // Navigate directly to the base URL
```
no comments should be presented in the code

MarkAtolog replied:
code should tell everything by himself. variable names, function names, they describe everything. If you need to leave a comment - your code is not understandable enough, rewrite it, make it easier to read and understand, it should be like a book

abubakar-a1qa committed:
Updated getItemNamesFromResultList() and waitForItemsToChange()

updated the method logic and timeout duration
